hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -502,8 +502,10 @@ public Cacheable getBlock(BlockCacheKey key, boolean caching, boolean repeat,
           // block will use the refCnt of bucketEntry, which means if two HFileBlock mapping to
           // the same BucketEntry, then all of the three will share the same refCnt.
           Cacheable cachedBlock = ioEngine.read(bucketEntry);
-          // RPC start to reference, so retain here.
-          cachedBlock.retain();
+          if (ioEngine.usesSharedMemory()) {
```

Comment:
Could we consider the patch in another side ?  Say  the BucketEntry#refCnt still means how many reference path from the RPC handler & BucketCache ... But the Cacheable#refCnt will always be 1 when the RPC is still handling.. Once the    cells shipped to client, then both the Cacheable#refCnt & BucketEntry#refCnt will decrease, actually the Cacheable will de-allocate the memory from ByteBuffAllocator. 
In this way, the BucketEntry#refCnt still have the same meaning as the SharedFileIOEngine in BucketCache. I mean  in BucketCache we don't need to care the IOEngine is shared or exclusive.

Reply:
let me think how to achieve this